### PR TITLE
Vector SNP genotyping spec

### PR DESCRIPTION
**Pull request by alimanfoo**
This PR brings in a pipeline spec fo SNP genotyping for vectors.

### CODE REVIEW

File: docs/specs/snp-genotyping-vector.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+* ``alt_number`` parameter should be 3 (fix number of alt alleles)
+
+* Data should be grouped by contig (chromosome). This means calling
+  ``vcf_to_zarr()`` once for each contig in the reference genome,
+  specifying the name of the contig as the value of the ``region`` and
+  ``group`` parameters.
+
+Here is an example of the appropriate function call in Python for a
+single contig (this should be called for all contigs in the genome):
+
```
Comment (tnguyensanger) [inline docs/specs/snp-genotyping-vector.md:129]: For reference, past pipelines used this script for converting from vcf to zarr:  https://github.com/malariagen/legacy_pipelines/blob/master/prod-tools/scripts/vcf_to_zarr.py

There is additional code for ensuring that the files within the zarr have group read/write permissions.  It's handy on the Sanger farm systems, but not necessary as long as the files are created by a production user.

File: docs/specs/snp-genotyping-vector.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    allel.vcf_to_zarr(
+        input=input_path,
+        output=output_path,
+        group=contig,
+        region=contig,
+        compressor=zarr.Blosc(cname='zstd', clevel=5, shuffle=0),
+        overwrite=True,
+        fields=['calldata/GT', 'calldata/GQ', 'calldata/AD', 'variants/MQ'],
+        types={'calldata/GT': 'i1',
+               'calldata/GQ': 'i1',
```
Comment (hardingnj) [inline docs/specs/snp-genotyping-vector.md:146]: I think explicitly setting `GQ` and `MQ` as integers is a change from the previous/current version and could be noted in the change log.

File: docs/specs/snp-genotyping-vector.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+### Step: Zip Zarr
+
+Once the zarr conversion is complete, create a zip archive from the
+contents of the root folder of the zarr directory. E.g., if
+``{sample}`` is the sample ID and ``{sample}.zarr`` is the name of the
+output directory used in calls to ``vcf_to_zarr()``, then do:
+
+```bash
+cd {sample}.zarr
+zip -rmT0 {sample}.zarr.zip .
```
Comment (hardingnj) [inline docs/specs/snp-genotyping-vector.md:166]: `-r` recurse paths
`-m` move, ie delete existing dir and contents
`-T` test, ie only delete on a successful zip. 
`-0` no compression.

File: docs/specs/snp-genotyping-vector.md (hunk 2 above)
Comment (hardingnj) [reply docs/specs/snp-genotyping-vector.md:146]: The theoretical maximum of `MQ` (from SAM format specs) is 255- therefore we need `i2`:
`GQ` is capped at 99, so `i1` is fine.

Also, might be worth noting here that MQ is the RMS mapping quality of all reads contributing to genotype call.

Comment (hardingnj) [reply docs/specs/snp-genotyping-vector.md:146]: Also, `GQ`/`MQ` can be missing. Unsure how `allel` handles this @alimanfoo ? Should be coded as `-1` in int output.

File: docs/specs/snp-genotyping-vector.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+* For all vector species we currently genotype at all sites in the
+  genome and all possible SNP alleles, which means that the set of
+  alleles provided to the genotyper comprises all genome sites where
+  the reference allele is not "N" and all possible SNP alternate
+  alleles are given in lexical order.
+
+
+## Change log
+
+* Version 1.4.0 - Updated Zarr and scikit-allel versions. Corrected
```
Comment (hardingnj) [inline docs/specs/snp-genotyping-vector.md:181]: `GQ` and `MQ` recorded as integers in the zarr output file.

File: docs/specs/snp-genotyping-vector.md (hunk 1 above)
Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:129]: Thanks Thuy. Just to note that as of zarr v2.4 it should no longer be necessary to change permissions, these will be set according to users umask.

File: docs/specs/snp-genotyping-vector.md (hunk 2 above)
Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:146]: > I think explicitly setting `GQ` and `MQ` as integers is a change from the previous/current version and could be noted in the change log.

The [previous version](https://github.com/malariagen/vector-ops/blob/a1227e1ae0dcdf950ecc633cb63c4da44711da66/specs/genotyping.md) did specify a dtype for GQ but was actually missing the MQ field completely, although we did include MQ in the implementation. I've added a note.

Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:146]: > The theoretical maximum of `MQ` (from SAM format specs) is 255- therefore we need `i2`:

I'm pretty sure that the maximum MQ from bwa is 60, so 'i1' should be fine.

Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:146]: > Also, `GQ`/`MQ` can be missing. Unsure how `allel` handles this @alimanfoo ? Should be coded as `-1` in int output.

Yes, all integer fields should get -1 if missing.

**Comment by alimanfoo** [comment]
Thanks @hardingnj for comments.

**Comment by alimanfoo** [comment]
I added an appendix with some documentation of the Zarr hierarchy structure we expect, thanks @kbergin for suggestion.

File: docs/specs/snp-genotyping-vector.md (hunk 2 above)
Comment (hardingnj) [reply docs/specs/snp-genotyping-vector.md:146]: Cool- just checked all phase 3 zip zarrs. `70` is the max observed.

Comment (tnguyensanger) [reply docs/specs/snp-genotyping-vector.md:146]: I believe the /variants/MQ field in the AG1K phase 3 zarrs were float32.  UnifiedGenotyper sets MQ annotation as the root mean square of alignment mapping quality as @hardingnj mentioned.  Just wanted to confirm - are we OK with losing the decimal precision for the /variants/MQ field in the zarrs going forward?

File: docs/specs/snp-genotyping-vector.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+For the output of the "VCF to Zarr" conversion step, we expect that
+this will create a Zarr hierarchy with the following structure:
+
+* `/` [root group]
+  * `{chromosome}` [group]
+    * `variants` [group]
+      * `MQ` [array; int8; shape (n_sites,)]
+    * `calldata` [group]
+      * `GT` [array; int8; shape (n_sites, 1, 2)]
```
Comment (tnguyensanger) [inline docs/specs/snp-genotyping-vector.md:189]: Could we can be more explicit on what each dimension represents?  That tripped me up when I first started to work with single-sample zarrs.  Maybe something like:

* `GT` [array; int8; shape (n_sites, samples=1 , ploidy=2)]

or 

* `GT` [array; int8; shape (n_sites, 1 , 2), dimensions correspond to (sites, samples, ploidy)]

Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:189]: Thanks, good suggestion.

File: docs/specs/snp-genotyping-vector.md (hunk 2 above)
Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:146]: > Cool- just checked all phase 3 zip zarrs. `70` is the max observed.

Thanks for doing that Nick.

Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:146]: > I believe the /variants/MQ field in the AG1K phase 3 zarrs were float32. UnifiedGenotyper sets MQ annotation as the root mean square of alignment mapping quality as @hardingnj mentioned. Just wanted to confirm - are we OK with losing the decimal precision for the /variants/MQ field in the zarrs going forward?

Ah OK. Yes should be fine as int8, definitely don't need the decimal precision.

Comment (tnguyensanger) [reply docs/specs/snp-genotyping-vector.md:146]: @alimanfoo @hardingnj  Sorry, I realize I left out some information.  To clarify, the MQ annotation is the root mean square **+ standard deviation**.  (According to:  https://gatk.broadinstitute.org/hc/en-us/articles/360037591751-RMSMappingQuality).  We'll lose information if we drop from float32 to int8.  Is this still OK?

Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:146]: Hi @tnguyensanger, bottom line is that range of values of MQ is 0-70, decimal precision is not needed for any analysis we do.

MQ is calculated as root mean square, which means square mapping qualities from all reads, sum the quares, divide by number of reads, take the root.

The "statistical note" in the page you linked says that RMS is "equivalent to the mean of the mapping qualities plus the standard deviation of the mapping qualities". That may be true, I haven't checked. In either case, it's just a point of statistical interest. Doesn't change the fact that we don't need the decimal precision.

Hth.

Comment (tnguyensanger) [reply docs/specs/snp-genotyping-vector.md:146]: Sounds good to me.

File: docs/specs/snp-genotyping-vector.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
+# SNP genotyping (vector) pipeline specification
+
+* Version: 1.4.0
```
Comment (gbggrant) [inline docs/specs/snp-genotyping-vector.md:3]: Starting out at 1.4.0?

File: docs/specs/snp-genotyping-vector.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+* Authors: Alistair Miles, Jim Stalker
+
+This document specifies a pipeline for genotyping an individual sample
+at a set of predefined SNP alleles, assuming the sample has already
+been aligned against the appropriate reference sequence via the short
+read alignment protocol.
```
Comment (gbggrant) [inline docs/specs/snp-genotyping-vector.md:9]: Maybe a reference to the spec here?

File: docs/specs/snp-genotyping-vector.md (hunk 6 above)
Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:3]: Slightly odd I know, but this carries on from the version numbering we had for this spec in the previous repo.

File: docs/specs/snp-genotyping-vector.md (hunk 7 above)
Comment (alimanfoo) [reply docs/specs/snp-genotyping-vector.md:9]: Will do.

**Comment by alimanfoo** [comment]
Thanks all for comments, going to merge this one but happy to follow up if further comments.

**Pull request merged**